lib/filewatch/tail_base.rb

```diff
@@ -52,183 +30,165 @@ def initialize(opts={})
-      elsif event == :create_initial
+        return true
+      elsif sdb_value && sdb_value.watched_file != watched_file
+        # we have seen this fingerprint before
+        # but it is allocated to a different file with the same (initial) content
+        # wf is a renamed file recently discovered
+        # or a different file with the same content
+        # to process this file we need a different key
+        # we will not open the file but we do deactivate it
+        # maybe one of the fingerprints will change.
+        @logger.debug? && @logger.debug("_add_to_sincedb: #{path}: found but differently allocated - setting wf back to watch for later retry")
+        watched_file.watch
+        return false
+      end
+
+      # if we get here, then we have a watched_file that has never been seen before.
+
+      sdb_value = SincedbValue.new(0)
+      sdb_value.upd_watched_file(watched_file) # <-- allocate this watched_file to the sincedb value
+      yield if block_given? # should open the file
+
+      case event
+      when :create_initial
         if @opts[:start_new_files_at] == :beginning
-          @logger.debug? && @logger.debug("#{path}: initial create, no sincedb, seeking to beginning of file")
+          @logger.debug? && @logger.debug("_add_to_sincedb: #{path}: initial create, no sincedb on disk, seeking to beginning of file")
           watched_file.file_seek(0)
-          @sincedb[sincedb_key] = 0
         else
           # seek to end
-          @logger.debug? && @logger.debug("#{path}: initial create, no sincedb, seeking to end #{stat.size}")
-          watched_file.file_seek(stat.size)
-          @sincedb[sincedb_key] = stat.size
+          @logger.debug? && @logger.debug("_add_to_sincedb: #{path}: initial create, no sincedb on disk, seeking to end #{stat.size}")
+          seek_to = stat.size
+          watched_file.file_seek(seek_to)
+          sdb_value.upd_position(seek_to)
         end
-      elsif event == :create
-        @sincedb[sincedb_key] = 0
-      elsif event == :modify && @sincedb[sincedb_key].nil?
-        @sincedb[sincedb_key] = 0
-      elsif event == :unignore
-        @sincedb[sincedb_key] = watched_file.bytes_read
+        @sincedb.set(sdb_key, sdb_value)
+      when :create, :shrink, :grow
+        @logger.debug? && @logger.debug("_add_to_sincedb: #{path}: #{event}, new content, no sincedb on disk, seeking to beginning of file")
+        watched_file.file_seek(0)
+        sdb_value.upd_position(0)
+        @sincedb.set(sdb_key, sdb_value)
+      when :unignore
+        # when this watched_file as ignored it had it bytes_read set to eof
+        watched_file.file_seek(watched_file.bytes_read)
+        sdb_value.upd_position(watched_file.bytes_read)
+        @sincedb.set(sdb_key, sdb_value)
       else
-        @logger.debug? && @logger.debug("#{path}: staying at position 0, no sincedb")
+        @logger.debug? && @logger.debug("_add_to_sincedb: #{path}: staying at position 0, no sincedb")
       end
```

Comment:
I see here that we have 2 case statements, I suggest we split the behavior in two separate methods, one for **new file found** and one for existing files.